### ISSUE TITLE
Updating symfony/console causing setup:di:compile failure

### ISSUE THREAD

**Issue by Skyfly2021** [magento/magento2#33595]
### Preconditions (*)

1.2.4.2-p1
2.Centos 8
3.php 7.4

### Steps to reproduce (*)

1.
2.

### Expected result (*)

1. [Screenshots, logs or description]
2.

### Actual result (*)

1. [Screenshots, logs or description]
2.Incompatible argument type: Required type: int. Actual type: string
The name of the upgrade package is not remembered
The following error occurred during compilation
![image](https://user-images.githubusercontent.com/53896477/127090358-bc07f4c6-0e74-4fa0-87cb-770955c23e8e.png)

### Workaround
The `symfony/console` released version 4.4.29 as a hotfix for Magento issue.
For now, you can run `composer update symfony/console`

Note: the root cause of this issue still not resolved yet

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by cha80** [comment]
![image](https://user-images.githubusercontent.com/56351661/127090712-7da9d41a-0ca7-49b1-b918-9de9d47b9f85.png)
 i am also face this problem 2.3.5-p2

**Comment by cha80** [comment]
it
might update some module case this problem

**Comment by Skyfly2021** [comment]
Please downgrade the package, finally solved：
1.composer require symfony/process:4.4.26
2.composer require symfony/console:4.4.26

**Comment by brosenberger** [comment]
i created a pull request for symfony/console: https://github.com/symfony/symfony/pull/42281 and also any 2.3.x magento2 version is affected

**Comment by cengizcoskun** [comment]
+1

**Comment by hostep** [comment]
Thanks @brosenberger & @Skyfly2021!

@Skyfly2021: I don't have issues with `symfony/process`, only with `symfony/console`

It would be great if Magento's `setup:di:compile` process could look at the actual argument types and only falls back to the ones in phpdoc when the argument doesn't have a type specified.

**Comment by kassner** [comment]
If you want to patch the issue instead of pinning symfony/console's version:

```
--- vendor/symfony/console/Input/InputOption.php
+++ vendor/symfony/console/Input/InputOption.php
@@ -48,6 +48,7 @@
     private $description;
 
     /**
+     * @param string $name
      * @param string|array|null                $shortcut The shortcuts, can be null, a string of shortcuts delimited by | or an array of shortcuts
      * @param int|null                         $mode     The option mode: One of the VALUE_* constants
      * @param string|bool|int|float|array|null $default  The default value (must be null for self::VALUE_NONE)
```

**Comment by mortenbirkelund** [comment]
Ran into this aswell.

**Comment by thomas-kl1** [comment]
It's not a symfony issue, it's Magento not being smart enough to resolve types correctly :)

Workaround:
"conflict": {
        "symfony/console": ">=v5.3.4||>=v4.4.27"
    },

**Comment by hostep** [comment]
I don't think that not allowing versions higher then or equal to 4.4.27 to be installed is the correct workaround. You tend to forget about these workarounds and if symfony then releases a newer version that fixes a security issue, you'll not get it.

It's probably better to only conflict with version 4.4.27 and still allow higher versions and then if it turns out that the next version still has the same problem, add that version to the conflict range.

It looks like they are [considering reverting the change](https://github.com/symfony/symfony/pull/42281#issuecomment-887605655) in the next version v4.4 of `symfony/console` (which should be version 4.4.28)

**Comment by thomas-kl1** [comment]
@hostep workaround is temporary fix, also the issue is not on the Symfony side, but Magento, as it was for the "Laminas issue" :)
I always encourage to update to the latest version of of trusted packages such as laminas or Symfony, because they are well tested, mature and robust. Most of time when there is an issue with a third party package upgrade, it's because Magento codebase is poorly implemented. Magento has an important debt that won't be fix because they know that "community" will find workaround or will fix it for them. Maybe we should let Magento assume their debt and pay it?

**Comment by viktym** [comment]
```
"conflict": {
    "symfony/console": "4.4.27 || 4.4.28"
  },
```
The changes will be reverted in symfony/console:4.4.29

**Comment by neha023** [comment]
There is a pull request for symfony/console. This will fix that issue.

https://github.com/symfony/console/pull/75

**Comment by fredden** [comment]
> This will fix that issue.

The issue exists within Magento. The change made in `symfony/console` is a temporary work-around and not a fix.

**Comment by thomas-kl1** [comment]
> > This will fix that issue.
> 
> The issue exists within Magento. The change made in `symfony/console` is a temporary work-around and not a fix.

![Thank you](https://media.giphy.com/media/5xtDarmwsuR9sDRObyU/giphy.gif)

**Comment by alainstout** [comment]
> Please downgrade the package, finally solved：
> 1.composer require symfony/process:4.4.26
> 2.composer require symfony/console:4.4.26

THIS: 2.composer require symfony/console:4.4.26 did the trick for me

Thank you!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @sidolov 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42929` were created

**Issue Available**: @sidolov, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ssx** [comment]
This also causes the same issue in `2.4.2-p2`

**Comment by barryvdh** [comment]
There is fix for the 4.4 version in Symfony/console, which isn't tagged yet. See https://github.com/symfony/symfony/pull/42281#issuecomment-887764179 for details.
Can you try that dev version to confirm no more changes are required?

This should still be fixed for the long term though.

**Comment by ihor-sviziev** [comment]
Hi,

I just did a short investigation - it looks like this issue is coming from the Laminas Code component.
Magento tries to detect type for constructor argument in the https://github.com/magento/magento2/blob/9756182a77a472be52ce995cb080ea742d3cd0ef/lib/internal/Magento/Framework/Code/Reader/ArgumentsReader.php#L111
while `$parameter->detectType()` is following:
https://github.com/laminas/laminas-code/blob/3.4.1/src/Reflection/ParameterReflection.php#L77-L113

For the `shortcut` argument on the `\Symfony\Component\Console\Input\InputOption::__construct`, it detects type - `int` (by `@param` tags position index, not by param name)

And, in the latest laminas-code, this issue isn't resolved yet as well.

**Comment by ihor-sviziev** [comment]
Note: symfony/console has released a new version - 4.4.29 where were resolved this issue.
https://github.com/symfony/console/compare/v4.4.27...v4.4.29

**Comment by ihor-sviziev** [comment]
FYI I prepare PR to the laminas-code to fix the listed issue.
https://github.com/laminas/laminas-code/pull/95
